src/fdTransfer.cpp

```diff
@@ -0,0 +1,416 @@
+        return false;
+    }
+
+    struct sockaddr_un sun;
+    socklen_t addrlen;
+    socketPathForPid(getpid(), &sun, &addrlen);
+
+    if (-1 == bind(_listener, (const struct sockaddr*)&sun, addrlen)) {
+        Log::error("FdTransfer bind(): %s", strerror(errno));
+        return false;
+    }
+
+    const struct timeval timeout = {
+        _accept_timeout_ms / 1000,
+        (_accept_timeout_ms % 1000) * 1000,
+    };
+    if (-1 == setsockopt(_listener, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout))) {
+        Log::error("FdTransfer setsockopt(SO_RCVTIMEO): %s", strerror(errno));
+        return false;
+    }
+
+    if (-1 == listen(_listener, 1)) {
+        Log::error("FdTransfer listen(): %s", strerror(errno));
+        return true;
+    }
+
+    return true;
+}
+
+bool FdTransfer::acceptPeer() {
+    _peer = accept(_listener, NULL, NULL);
+    if (_peer == -1) {
+        Log::error("FdTransfer accept(): %s", strerror(errno));
+        return false;
+    }
+
+    return true;
+}
+
+int FdTransfer::requestPerfFd(pid_t tid, struct perf_event_attr *attr) {
+    struct perf_fd_request request;
+
+    request.header.type = PERF_FD;
+    request.header.length = sizeof(request);
+    request.header.request_id = nextRequestId();
+
+    request.tid = tid;
+    memcpy(&request.attr, attr, sizeof(request.attr));
+
+    bool eof;
+    if (!send_all(_peer, &request, sizeof(request), &eof)) {
+        if (eof) {
+            Log::warn("FdTransfer unexpected EOF");
+        } else {
+            Log::warn("FdTransfer send(): %s", strerror(errno));
+        }
+        return -1;
+    }
+
+    return recvFd(request.header.request_id);
+}
+
+int FdTransfer::requestKallsymsFd() {
+    struct fd_request request;
+
+    request.type = KALLSYMS_FD;
+    request.length = sizeof(request);
+    request.request_id = nextRequestId();
+
+    bool eof;
+    if (!send_all(_peer, &request, sizeof(request), &eof)) {
+        if (eof) {
+            Log::warn("FdTransfer unexpected EOF");
+        } else {
+            Log::warn("FdTransfer send(): %s", strerror(errno));
+        }
+        return -1;
+    }
+
+    return recvFd(request.request_id);
+}
+
+// about perror usage, see comment on serveRequests
+bool FdTransfer::sendFd(int fd, unsigned int request_id) {
+    struct msghdr msg = {0};
+
+    struct iovec iov[1] = {
+        {
+            &request_id,
+            sizeof(request_id),
+        },
+    };
+    msg.msg_iov = iov;
+    msg.msg_iovlen = ARRAY_SIZE(iov);
+
+    union {
+       char buf[CMSG_SPACE(sizeof(fd))];
+       struct cmsghdr align;
+    } u;
+
+    if (fd != -1) {
+        msg.msg_control = u.buf;
+        msg.msg_controllen = sizeof(u.buf);
+
+        struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+        cmsg->cmsg_level = SOL_SOCKET;
+        cmsg->cmsg_type = SCM_RIGHTS;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(fd));
+        memcpy(CMSG_DATA(cmsg), &fd, sizeof(fd));
+    }
+
+    ssize_t ret = sendmsg(_peer, &msg, 0);
```

Comment:
After `fd` has been sent, the original `fd` needs to be closed.

Reply:
Mhm. Generally I paid less attention in the program side, because I had in mind that it's short-lived (not necessarily correct in long profiling sessions). Will fix this & do another pass on fd leaks etc in the program side.

Reply:
Done